atomic_reactor/plugins/post_koji_import.py

```diff
@@ -613,20 +564,126 @@ def _filesystem_koji_task_id(self) -> Optional[int]:
+            return
+
+        metadata = get_output_metadata(operator_manifests_path,
+                                       OPERATOR_MANIFESTS_ARCHIVE)
+        metadata['buildroot_id'] = buildroot_id
+        add_type_info(metadata, KOJI_BTYPE_OPERATOR_MANIFESTS)
+        outputs_metadata.append(metadata)
+
+        wf_data.koji_upload_files.append({
+            'local_filename': operator_manifests_path,
+            'dest_filename': OPERATOR_MANIFESTS_ARCHIVE,
+        })
 
+    def _collect_maven_metadata(
+            self, buildroot_id: str, outputs_metadata: List[Dict[str, Any]],
+    ) -> None:
+        wf_data = self.workflow.data
+        result = wf_data.postbuild_results.get(PLUGIN_MAVEN_URL_SOURCES_METADATA_KEY) or {}
+        for remote_source_file in result.get('remote_source_files', []):
+            metadata = remote_source_file['metadata']
+            metadata['buildroot_id'] = buildroot_id
+            outputs_metadata.append(metadata)
+            wf_data.koji_upload_files.append({
+                'local_filename': remote_source_file['file'],
+                'dest_filename': metadata['filename'],
+            })
+
+    def get_output(self, buildroot_id: str) -> List[Dict[str, Any]]:
+        """Assemble outputs specific to a binary build.
+
+        The corresponding files to be uploaded are also recorded for later
+        upload.
+
+        :param str buildroot_id: for binary build, this argument is ignored.
+            Instead, use the buildroot id which is already set in the build
+            metadata.
         :return: list, containing dicts of partial metadata
         """
-        outputs = []
-        output_file = None
+        wf_data = self.workflow.data
 
-        for platform, instance in self._iter_build_metadata_outputs():
-            instance['buildroot_id'] = '{}-{}'.format(platform, instance['buildroot_id'])
-            outputs.append(instance)
+        result = wf_data.prebuild_results.get(PLUGIN_FETCH_MAVEN_KEY) or {}
+        maven_components = result.get('components', [])
 
-        return outputs, output_file
+        platform: str
+        output: Dict[str, Any]  # an output metadata of the build
+        outputs: List[Dict[str, Any]] = []
+
+        for platform, output in self._iter_build_metadata_outputs():
+            buildroot_id = output['buildroot_id']
+            output['buildroot_id'] = f'{platform}-{buildroot_id}'
+            if maven_components and output['type'] == 'docker-image':
+                # add maven components alongside RPM components
+                output['components'] += maven_components
+            outputs.append(output)
+
+        buildroot_id = outputs[0]['buildroot_id']
+        self._collect_exported_operator_manifests_output(buildroot_id, outputs)
+        self._collect_source_tarballs_outputs(buildroot_id, outputs)
+        self._collect_remote_sources_json_outputs(buildroot_id, outputs)
+        self._collect_maven_metadata(buildroot_id, outputs)
```

Comment:
The `collect_*` methods all have a lot of things in common:
* they append to the `outputs` param
* they append to `self.workflow.data.koji_upload_files`
* they set the `buildroot_id` on outputs

Is there a reason to do it in those methods instead of doing it here? Something like
```python
platform_independent_outputs = [
    *self._collect_a(), *self._collect_b(), ...
]
for output in platform_indepent_outputs:
    # set buildroot_id, append to koji_upload_files, ...
```

Reply:
One major reason is trying to keep the caller side less code as much as possible. Meanwhile, for the individual collect method, it is also natural to express the meaning, a collect method will collect output metadata and corresponding files into proper place for later operation. As a result, there is no such a cycle, calling a collect method, the method returns something, caller operates the return value.

Reply:
It was actually a little painful for me when I was handling the outputs related methods. I thought it would be nice to make the collection action like a stream, from where to get the each output metadata and associated file. But, didn't figure out a good way yet.

Reply:
Ok, seems reasonable